website/scripts/ros/ros_traffic_signal.js

```diff
@@ -30,119 +30,239 @@ function TrafficSignalInfoList(){
+        } catch(error) {
             console.error(error);
-         }  
+        }
 
-         //set back to black after 1 second.
-        setInterval(function()
-        {
-            let elapsed_time = Date.now() - latest_start_time;
-            if(elapsed_time >= 1000)
-            {
+        // Create only one interval to avoid memory leaks
+        if (!window.signalIntervalCreated) {
+            window.signalIntervalCreated = true;
+            setInterval(function() {
+                let elapsed_time;
+
+                // Use sim clock for timing if available
+                if (useSimClock && simClockTime !== null && sim_start_time !== null) {
+                    // Calculate elapsed time in simulation
+                    let currentRealTime = Date.now();
+                    let realTimeDiff = currentRealTime - lastRealTime;
+
+                    // Estimate current sim time based on real time elapsed and last known sim time
+                    // This handles the case where sim_clock messages come at lower frequency than needed
+                    let estimatedSimTime = lastSimTime + realTimeDiff;
+                    elapsed_time = estimatedSimTime - sim_start_time;
+
+                } else {
+                    // Fall back to real clock
+                    elapsed_time = Date.now() - latest_start_time;
+                }
+
+                if (elapsed_time >= 1000) {
                     $('.traffic-signal-col').html(updateTrafficSignal('',''));
                     signalState = null;
-            }
-         }, 1000);
+                }
+            }, 1000);
+        }
     });
 }
 
 /****
  * @brief Given the timing from the icoming spat message, and return the count down for traffic signal phase change
  * @param Input moy (unit of min): Min of current UTC year.
  *              min_end_time (unit of 1/10 sec): Timing data in UTC time stamps for event. Include min end times of phase confidentce and estimated next occurrence
+ *              useSimClock: Boolean indicating whether to use simulation clock
+ *              simClockTime: Current simulation time in milliseconds (if available)
  * @returns current_phase_max_sec: The traffic signal will change phase in number of seconds.
  */
-function getCurPhaseMaxSecBySpatTiming(moy, min_end_time )
-{
-    let current_phase_max_sec = 0; 
+function getCurPhaseMaxSecBySpatTiming(moy, min_end_time, useSimClock = false, simClockTime = null) {
+
+    // Check for invalid inputs that could cause NaN
+    if (moy === undefined || moy === null || isNaN(moy)) {
+        console.error("[ERROR] Invalid moy value:", moy);
+        return 30; // Default fallback
+    }
+
+    if (min_end_time === undefined || min_end_time === null || isNaN(min_end_time)) {
+        console.error("[ERROR] Invalid min_end_time value:", min_end_time);
+        return 30; // Default fallback
+    }
+
+    let current_phase_max_sec = 0;
+
+    //get current year
+    let current_date;
```

Comment:
Maybe initialize with new Date(now) and update it when using Sim clock.

Reply:
Done